proposals/NNNN-retroactive-protocol-refinement.md

```diff
@@ -0,0 +1,100 @@
+
+Swift-evolution thread: [Protocol extensions inheriting protocols](https://forums.swift.org/t/protocol-extensions-inheriting-protocols/25491/8)
+
+## Motivation
+
+Extensions to classes or structs (nominals) are able to specify conformances that they implement. When one extends a protocol however, it seems reasonable that specifying a conformance would add that functionality to that an entire class of nominals i.e. those that already adopt the protocol being extended or protocols that adopt that protocol etc. One example that was discussed was something that should be a simple case, where one wants to make all fixed width integers expressible by a UnicodeScalarLiteral (i.e. Strings). On the face of it one might assume the following would be possible:
+
+```Swift
+extension FixedWidthInteger: ExpressibleByUnicodeScalarLiteral {
+  @_transparent
+  public init(unicodeScalarLiteral value: Unicode.Scalar) {
+    self = Self(value.value)
+  }
+}
+```
+In practical terms, for the Swift compiler, this effectively means creating ad-hoc extensions on Int8, Int16 and all types which conform to FixedWidthInteger along with their associated witnesses. It's more powerful however as an idea and in more abstract terms (from the Generics Manifesto):
+
+```
+protocol P {
+  func foo()
+}
+
+protocol Q {
+  func bar()
+}
+
+extension Q : P { // Make every type that conforms to Q also conforms to P
+  func foo() {    // Implement `P.foo` requirement in terms of `Q.bar`
+    bar()
+  }
+}
+
+func f<T: P>(t: T) { ... }
+
+struct X : Q {
+  func bar() { ... }
+}
+
+f(X()) // okay: X conforms to P through the conformance of Q to P
+```
+```
+This is an extremely powerful feature: it allows one
+to map the abstractions of one domain into another
+domain (e.g., every Matrix is a Graph).
+```
+It continues though:
+
+```
+However, similar to private conformances, it puts a
+major burden on the dynamic-casting runtime to chase
+down arbitrarily long and potentially cyclic chains
+of conformances, which makes efficient implementation
+nearly impossible.
+```
+The PR above strives to demonstrate that it may be possible to avoid these pitfalls and refine protocols retroactively at compile time, providing summary conformance data to the existing runtime for an efficient implementation of dynamic casts.
+
+## Proposed solution
+
+Allow conformances to be added to protocol extensions with all that implies in terms of all nominals conforming to the protocol being extended acquiring the added conformances. It would be useful if this also worked for synthesised conformances (and indeed it does).
```

Comment:
```suggestion
Allow conformances to be added to protocol extensions with all that implies in terms of all nominals conforming to the protocol being extended acquiring the added conformances. It would be useful if this also worked for synthesized conformances (and indeed it does).
```